pennylane/templates/subroutines/amplitude_amplification.py

```diff
@@ -0,0 +1,176 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This submodule contains the template for Amplitude Amplification.
+"""
+
+# pylint: disable-msg=too-many-arguments
+import numpy as np
+from pennylane.operation import Operation
+import pennylane as qml
+
+
+def _get_fixed_point_angles(iters, p_min):
+    """
+    Returns the angles needed for the fixed-point amplitude amplification algorithm.
+    The angles are computed using equation (11) of  `arXiv:1409.3305v2 <https://arxiv.org/abs/1409.3305>`__.
+    """
+
+    delta = np.sqrt(1 - p_min)
+    gamma = np.cos(np.arccos(1 / delta, dtype=np.complex128) / iters, dtype=np.complex128) ** -1
+
+    alphas = [
+        2 * np.arctan(1 / (np.tan(2 * np.pi * j / iters) * np.sqrt(1 - gamma**2)))
+        for j in range(1, iters // 2 + 1)
+    ]
+    betas = [-alphas[-j] for j in range(1, iters // 2 + 1)]
+    return alphas[: iters // 2], betas[: iters // 2]
+
+
+class AmplitudeAmplification(Operation):
+    r"""Applies amplitude amplification.
+
+    Given a state :math:`|\Psi\rangle = \alpha |\phi\rangle + \beta|\phi^{\perp}\rangle`, this subroutine amplifies the amplitude of the state :math:`|\phi\rangle` such that
+
+    .. math::
+
+            \text{A}(U, O)|\Psi\rangle \sim |\phi\rangle.
+
+    The implementation of the algorithm is based on `[arXiv:quant-ph/0005055] <https://arxiv.org/abs/quant-ph/0005055>`__.
+    The template also unlocks advanced techniques such as fixed-point quantum search [`arXiv:1409.3305 <https://arxiv.org/abs/1409.3305>`__] and oblivious amplitude amplification [`arXiv:1312.1414 <https://arxiv.org/abs/1312.1414>`__] by reflecting on a subset of the wires.
+
+    Args:
+        U (Operator): operator that prepares the state :math:`|\Psi\rangle`
+        O (Operator): the oracle that flips the sign of the state :math:`|\phi\rangle` and does nothing to the state :math:`|\phi^{\perp}\rangle`
+        iters (int): the number of iterations of the amplitude amplification subroutine, default is ``1``
+        fixed_point (bool): whether to use the fixed-point amplitude amplification algorithm, default is ``False``
+        work_wire (int): the auxiliary wire to use for the fixed-point amplitude amplification algorithm, default is ``None``
+        reflection_wires (Wires): the wires to reflect on, default is the wires of ``U``
+        p_min (int): the lower bound for the probability of success in fixed-point amplitude amplification, default is ``0.9``
+
+    Raises:
+        ValueError: ``work_wire`` must be specified if ``fixed_point == True``.
+        ValueError: ``work_wire`` must be different from the wires of the oracle ``O``.
+
+    **Example**
+
+    Amplification of state :math:`|2\rangle` using Grover's algorithm with 3 qubits.
+    The state :math:`|\Psi\rangle` is constructed as a uniform superposition of basis states.
+
+    .. code-block::
+
+        @qml.prod
```

Comment:
Please use standard indentation here as well.

Reply:
my tab does strange things to me. Fixed 👍